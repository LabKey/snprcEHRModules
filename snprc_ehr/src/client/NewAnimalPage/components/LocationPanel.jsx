import React from 'react';
import Select from 'react-select';
import WrappedDatePicker from '../../utils/components/WrappedDatePicker';

export default class LocationPanel extends React.Component {

    state = {
        errorMessage: undefined
    };

    componentDidMount = () => {
        this.props.preventNext(!this.props.newAnimalData.room); //prevent/Allow Next button
    }
    handleCageChange = e => {
        const value = e.target.value;
        const validatedOption = this.validate('cage', { value });
        if (!validatedOption.hasError) {
            this.props.handleDataChange('cage', validatedOption);
        }
        
    }

    handleRoomChange = room => {
        const el = document.getElementById("cage-input");
        if (room.maxCages) {
            el.disabled = false
            el.value = null;
            el.max = room.maxCages;
        }
        else {
            el.disabled = true;
            el.value = null
        }
        this.props.handleDataChange('room', this.validate('room', room));
        this.props.handleDataChange('cage', {value: null, hasError: false} );
        this.props.preventNext(false); //prevent/Allow Next button
    }

    validate = (property, option) => {
        switch (property) {
            case 'cage':
                const { room } = this.props.newAnimalData;
                let hasError = false;
                if (room && room.maxCages) {
                   
                    hasError = option.value !== ""  && (option.value > room.maxCages || option.value < 1);

                    const errorMessage = hasError ? `Cage # must be between 1 and ${room.maxCages} in ${room.value}` : undefined;

                    this.setState((prevState) => (
                        {
                            ...prevState,
                            errorMessage: errorMessage
                        }
                    ));
                }

                this.props.handleError(hasError);
                return ({
                    ...option,
                    hasError
                })
                

            case 'room':
                // TODO: what constitutes a room error?
                return ({
                    ...option,
                    hasError: false
                })
        }
    }
    handleDateSelect = date => {
        //do nothing
    };
    handleDateChangeRaw = e => {
        e.preventDefault();
    }
    isInteger = e => {
        const i = e.key    //which ? e.which : e.keyCode;
        console.log(i);
        const isInteger =  (i >= 0 && i <= 9);
        if (!isInteger) {
            e.preventDefault();
        }

        return isInteger;
    }  

    render() {
        let { room, cage, acqDate } = this.props.newAnimalData;
        return (
            <div className="wizard-panel__rows">
                <div className="wizard-panel__row" >
                    <div className="wizard-panel__col">
                        <div className="location-datepicker">
                            <WrappedDatePicker
                                label="Move Date Time"
                                todayButton="Today"
                                dateFormat="Pp"
                                selected={acqDate.date}
                                onSelect={this.handleDateSelect}
                                onChange={this.handleDateSelect}
                                onChangeRaw={this.handleDateChangeRaw}
                                readOnly={true}
                                disabledKeyboardNavigation={true}
                                disabled={this.props.disabled}
                            />
                        </div>
                    </div>
                </div>
                <div className="wizard-panel__row" >
                    <div className="wizard-panel__col">
                        <label className="field-label" >Location</label>
                        <Select
                            defaultValue={room}
                            className="shared-dropdown"
                            classNamePrefix="shared-select"
                            options={this.props.locationList}
                            onChange={this.handleRoomChange}
                            placeholder="Select Location"
                            isDisabled={this.props.disabled}
                            id="location-select"
                        />
                    </div>
                    <div className="wizard-panel__col">
                        <label className="field-label" >Cage</label>
                        <input type="number" 
                            onKeyPress={this.isInteger}
                            id="cage-input"
                            className="cage-input"
                            placeholder="Cage #"
                            min="1"
                            max={room ? room.maxCages : 1}
                            defaultValue={cage.value}
                            onChange={this.handleCageChange}
                            disabled={!(room && room.maxCages)}

                        />
                    </div>
                </div>
                <div className="wizard-panel__row" >
                    <div className="err-panel">
                        {!room && 'You must select a location.'}
                        {this.state.errorMessage && this.state.errorMessage}
                    </div>
                </div>
            </div>

        )
    }
}
