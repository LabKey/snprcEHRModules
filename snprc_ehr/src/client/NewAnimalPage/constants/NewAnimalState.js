import moment from 'moment';

const NewAnimalState = () => {

    return  ({
        currentPanel: 1,
        selectedOption: undefined,
        newAnimalData: {
            selectedOption: undefined,
            id: undefined,
            birthDate: { date: moment() },
            bdStatus: undefined,
            acquisitionType: undefined,
            acqDate: { date: moment() },
            gender: undefined,
            sire: undefined,
            dam: undefined,
            species: undefined,
            colony: undefined,
            animalAccount: undefined,
            ownerInstitution: undefined,
            responsibleInstitution: undefined,
            room: undefined,
            cage: { value: undefined },
            diet: undefined,
            pedigree: undefined,
            iacuc: undefined
        },
        summaryData: [],
        speciesList: [],
        acquisitionTypeList: [],
        potentialDamList: [],
        potentialSireList: [],
        locationList: [],
        accountList: [],
        colonyList: [],
        dietList: [],
        institutionList: [],
        iacucList: [],
        pedigreeList: [],
        bdStatusList: [],
        isLoading: true,
        hasError: false,
        preventNext: true,
        saveOk: false,
        showSaveModal: false,
        showCancelModal: false,
        showSpeciesChangeModal: false,
        errorMessage: undefined
    });
}

export default NewAnimalState;
