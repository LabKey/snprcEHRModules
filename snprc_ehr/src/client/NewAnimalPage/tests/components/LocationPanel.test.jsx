import React from 'react'
import { shallow } from 'enzyme'
import LocationPanel from '../../components/LocationPanel'
import NewAnimalState from '../../constants/NewAnimalState'
import lists from '../fixtures/lists'

jest.mock('moment', () => {
    const moment = jest.requireActual('moment')('2023-01-01T00:00:00.000Z');
    return () => moment.utc(0);
});

test('Should render the LocationPanel', () => {
    const { newAnimalData } = new NewAnimalState()

    const wrapper = shallow(
      <LocationPanel
        disabled={ false }
        handleDataChange={ () => { return false } }
        handleError={ () => { return false } }
        locationList={ lists.locationList }
        newAnimalData={ newAnimalData }
        preventNext={ () => { return false } }
      />
    )
    expect(wrapper).toMatchSnapshot()
})
