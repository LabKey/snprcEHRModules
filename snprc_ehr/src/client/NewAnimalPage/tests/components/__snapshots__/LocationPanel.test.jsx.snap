// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the LocationPanel 1`] = `
<Fragment>
  <div
    className="wizard-panel__rows"
  >
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <div
          className="location-datepicker"
        >
          <WrappedDatePicker
            dateFormat="Pp"
            disabled={false}
            disabledKeyboardNavigation={true}
            label="Move Date Time"
            onChange={[Function]}
            onChangeRaw={[Function]}
            onSelect={[Function]}
            readOnly={true}
            selected={1970-01-01T00:00:00.000Z}
            todayButton="Today"
          />
        </div>
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Location
        </label>
        <StateManager
          autoFocus={true}
          className="shared-dropdown"
          classNamePrefix="shared-select"
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          id="location-select"
          isDisabled={false}
          isLoading={false}
          onChange={[Function]}
          options={
            Array [
              Object {
                "arcSpeciesCode": "PC",
                "id": 0,
                "label": "1.01",
                "maxCages": 42,
                "value": "1.01",
              },
              Object {
                "arcSpeciesCode": "PC",
                "id": 1,
                "label": "1.02",
                "maxCages": 43,
                "value": "1.02",
              },
            ]
          }
          placeholder="Select Location"
        />
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Cage
        </label>
        <input
          className="cage-input"
          disabled={true}
          id="cage-input"
          max={1}
          min="1"
          onChange={[Function]}
          onKeyPress={[Function]}
          onPasteCapture={[Function]}
          placeholder="Cage #"
          type="number"
        />
      </div>
    </div>
  </div>
  <InfoPanel
    messages={
      Array [
        Object {
          "colName": "Location",
          "propTest": true,
        },
      ]
    }
  />
</Fragment>
`;
